File: app/models/Component.scala
package models

import models.LineChart.Datum
import play.api.libs.json.{JsArray, JsBoolean, JsNull, JsNumber, JsObject, JsString, JsValue}

trait Component[T <: Any] {
  val cls: String
  val name: String
  val value: T
}

object Component {
  def importFromJson(value: JsValue): Component[_] = {
    val obj = value.as[JsObject].value

    val componentName = obj("name").as[JsString].value
    obj("cls") match {
      case JsString("Variable") =>
        val name = componentName
        val value = obj("value") match {
          case JsNumber(value)  => value.toDouble
          case JsString(value)  => value
          case JsBoolean(value) => value
          case JsNull           => null
          case _                => throw new Exception("Unknown type")
        }
        Variable(name, value)
      case JsString("ProgressBar") => ProgressBar(
          componentName,
          obj("value").as[JsNumber].value.toInt,
          obj("total").as[JsNumber].value.toInt
        )
      case JsString("Image") => Image(
          componentName,
          obj("value").as[JsString].value
        )
      case JsString("Logger") => Logger(
          componentName,
          obj("value").as[JsArray].value.map(_.as[JsString].value).toList
        )
      case JsString("LineChart") => LineChart(
          componentName,
          obj("value")
            .as[JsArray]
            .value
            .map(datum => {
              val datumObj = datum.as[JsObject].value
              LineChart.Datum(
                datumObj("x").as[JsNumber].value.toDouble,
                datumObj("y").as[JsNumber].value.toDouble,
                datumObj("label").as[JsString].value
              )
            })
            .toList,
          obj("x_label").as[JsString].value,
          obj("y_label").as[JsString].value
        )
      case _ => throw new Exception("Unknown type")
    }
  }
}

case class Variable[T](name: String, value: T) extends Component[T] {
  override val cls: String = "Variable"
}

case class ProgressBar(name: String, value: Int, total: Int) extends Component[Int] {
  override val cls: String = "ProgressBar"
}

case class Image(name: String, value: String /*image path*/ ) extends Component[String] {
  override val cls: String = "Image"
}

case class Logger(name: String, value: List[String] /*message rows*/ ) extends Component[List[String]] {
  override val cls: String = "Logger"
}

case class LineChart(
    name: String,
    value: List[Datum],
    xLabel: String,
    yLabel: String
) extends Component[List[Datum]] {
  override val cls: String = "LineChart"
}

object LineChart {
  case class Datum(x: Double, y: Double, label: String)
}

File: client/python/src/mita_client/component.py
from abc import ABC
from dataclasses import dataclass, asdict
from typing import Generic, TypeVar, Union, List, Dict

T = TypeVar("T")


@dataclass
class Component(ABC, Generic[T]):
    cls: str
    name: str
    value: T

    def to_dict(self) -> dict:
        return asdict(self)


class Variable(Component[T]):

    def __init__(self, name: str, value: T):
        super().__init__("Variable", name, value)

    def set(self, value: T):
        self.value = value


class ProgressBar(Component[Union[int, float]]):

    def __init__(self, name: str, value: Union[int, float] = 0, total: Union[int, float] = 100):
        super().__init__("ProgressBar", name, value)
        self.total = total

    def to_dict(self) -> dict:
        return {**super().to_dict(), "total": self.total}

    def set(self, value: Union[int, float]):
        self.value = value


class Logger(Component[List[str]]):

    def __init__(self, name: str):
        super().__init__("Logger", name, [])

    def log(self, value: str):
        self.value.append(value)

    def to_dict(self) -> dict:
        return {**super().to_dict()}


class LineChart(Component[List[Dict[str, Union[float, str]]]]):

    def __init__(self, name: str, x_label: str = "index", y_label: str = "value"):
        super().__init__("LineChart", name, [])
        self.x_label = x_label
        self.y_label = y_label

    def add(self, x: float, y: float, label: str = ""):
        self.value.append({"x": x, "y": y, "label": label})

    def to_dict(self) -> dict:
        return {**super().to_dict(), "x_label": self.x_label, "y_label": self.y_label}
